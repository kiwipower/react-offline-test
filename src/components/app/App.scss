.App {
  font-size: 16px;

  h1 {
    text-align: center;
    color: #4B0082;
  }

  .power-sources {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
  }

  @keyframes lds-hourglass {
    0% {
      transform: rotate(0);
      animation-timing-function: cubic-bezier(0.55, 0.055, 0.675, 0.19);
    }
    50% {
      transform: rotate(900deg);
      animation-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
    }
    100% {
      transform: rotate(1800deg);
    }
  }

}

.loading {
  display: inline-block;
  position: absolute;
  left: calc(50% - 32px);
  top: calc(50% - 32px);
  width: 64px;
  height: 64px;
}

.loading:after {
  content: " ";
  display: block;
  border-radius: 50%;
  width: 0;
  height: 0;
  margin: 6px;
  box-sizing: border-box;
  border: 26px solid #FF0000;
  border-color: #FF0000 transparent #FF0000 transparent;
  animation: lds-hourglass 1.2s infinite;
}
