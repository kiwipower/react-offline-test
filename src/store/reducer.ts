import { ApplicationState, ApplicationAction } from './types';

export const initialState: ApplicationState = {
    loading: {
        productionTypes: false,
    },
    productionTypes: [],
};

const reducer = (state = initialState, action: ApplicationAction) => {
    switch (action.type) {
        case 'loadProductionTypesRequest':
            return {
                ...state,
                loading: {
                    productionTypes: true,
                }
            };

        case 'loadProductionTypesSuccess':
            return {
                ...state,
                productionTypes: action.productionTypes,
                loading: {
                    productionTypes: false
                }
            };

        case 'loadProductionTypesError':
            return {
                ...state,
                loading: {
                    productionTypes: false
                }
            };

        default:
            return {
                ...state
            }
    }
};

export default reducer;
