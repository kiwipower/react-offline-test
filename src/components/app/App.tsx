import * as React from 'react';
import { connect } from 'react-redux';

import { loadProductionTypes } from '../../store/effects'
import { ApplicationState, ProductionType } from "../../store/types";
import { PowerSource } from "../power-source/PowerSource";
import { Chart } from "../chart/Chart";
import './App.scss';

interface Props {
    dispatch,
    loading,
    productionTypes: ProductionType[]
}

class App extends React.Component<Props, ApplicationState> {

    componentDidMount() {
        const { dispatch } = this.props;
        dispatch(loadProductionTypes());
    }

    render() {
        const { productionTypes, loading } = this.props;

        if (loading.productionTypes) {
            return <div className="loading" />
        }

        return (
            <div className="App">
                <h1>UK Energy Mix</h1>
                <div className="power-sources">
                    {
                        productionTypes && productionTypes.map ((productionType)=> {
                            return <PowerSource key={productionType.fuel} fuel={productionType.fuel} perc={productionType.perc} />
                        })
                    }
                </div>
                <Chart productionTypes={productionTypes}/>
            </div>
        );
    }
}

const mapStateToProps = ( state: ApplicationState ) => {
    const { productionTypes, loading } = state;
    return {
        productionTypes,
        loading
    }
};

export default connect( mapStateToProps )( App )
