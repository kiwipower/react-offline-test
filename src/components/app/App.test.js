import React from 'react';
import { shallow } from 'enzyme';
import { composeWithDevTools } from "redux-devtools-extension";
import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";

import App from './App';
import reducer, {initialState} from "../../store/reducer";

const productionTypes =  [
  { fuel: 'biomass', perc: 3.6 },
  { fuel: 'wind', perc: 6 },
  { fuel: 'solar', perc: 0.3 },
];

const composeEnhancers = composeWithDevTools({});
const store = createStore(reducer, initialState, composeEnhancers(applyMiddleware(thunk)));

const defaultProps = { productionTypes, store };

describe('App', () => {
  test('should render correctly', () => {
    const component = shallow(<App {...defaultProps}/>);
    expect(component).toMatchSnapshot();
  });
});
