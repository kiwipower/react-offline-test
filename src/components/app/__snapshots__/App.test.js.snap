// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <App
    dispatch={[Function]}
    loading={
      Object {
        "productionTypes": false,
      }
    }
    productionTypes={Array []}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</ContextProvider>
`;
